### PR TITLE
Tolerate overaligned MIR constants for codegen.

### LINKED ISSUE #117761: Enabling GVN produces a const allocation with the wrong alignment?
```
rustc +nightly ice.rs -Copt-level=0 -Zmir-opt-level=2 -Zinline-mir -Zmir-enable-passes=+GVN
```
```rust
struct Affine2 {
    matrix2: Mat2,
    translation: Vec2,
}

impl Affine2 {
    #[inline]
    fn inverse(&self) {
        mat2_ref(&self.matrix2);
        vec2_move(self.translation);
    }
}

#[derive(Clone, Copy)]
#[repr(align(8))]
struct Mat2([f32; 4]);

#[derive(Clone, Copy)]
struct Vec2 {
    x: f32, 
    y: f32, 
}

#[inline(never)]
fn mat2_ref(_: &Mat2) {
    loop {}
}

#[inline(never)]
fn vec2_move(_: Vec2) {
    loop {}
}

fn main() {
    Affine2 {
        matrix2: Mat2([0.0, 0.0, 0.0, 0.0]),
        translation: Vec2 { x: 0.0, y: 0.0 },
    }.inverse();
}
```

Error:
```
thread 'rustc' panicked at /rustc/f967532a47eb728ada44473a5c4c2eca1a45fe30/compiler/rustc_codegen_ssa/src/mir/operand.rs:135:9:
assertion `left == right` failed
  left: Align(8 bytes)
 right: Align(4 bytes)
```
That's this:
https://github.com/rust-lang/rust/blob/06e02d5b259c1e88cbf0c74366d9e0a4c7cfd6d9/compiler/rustc_codegen_ssa/src/mir/operand.rs#L128-L135

The GVN MIR diff is this:
```diff
diff --git a/ice.main.005-017.GVN.before.mir b/ice.main.005-017.GVN.after.mir
index 512091e..af00cdf 100644
--- a/ice.main.005-017.GVN.before.mir
+++ b/ice.main.005-017.GVN.after.mir
@@ -1,4 +1,4 @@
-// MIR for `main` before GVN
+// MIR for `main` after GVN
 
 fn main() -> () {
     let mut _0: ();
@@ -22,7 +22,12 @@ fn main() -> () {
     }
 
     bb2: {
-        _5 = ((*_1).1: Vec2);
-        _4 = vec2_move(move _5) -> [return: bb1, unwind continue];
+        _5 = const Vec2 {{ x: 0f32, y: 0f32 }};
+        _4 = vec2_move(const Vec2 {{ x: 0f32, y: 0f32 }}) -> [return: bb1, unwind continue];
     }
 }
+
+alloc11 (size: 24, align: 8) {
+    0x00 │ 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 │ ................
+    0x10 │ 00 00 00 00 00 00 00 00                         │ ........
+}
```

searched nightlies: from nightly-2023-06-01 to nightly-2023-11-08
regressed nightly: nightly-2023-10-30
searched commit range: https://github.com/rust-lang/rust/compare/e5cfc55477eceed1317a02189fdf77a4a98f2124...608e9682f0a6482903de8d3332770104a0ad943c
regressed commit: https://github.com/rust-lang/rust/commit/83c9732e0c7e1cae5f039677da4c51ee1d9b19b0

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2023-06-01 --end=2023-11-08 --script script 
```
</details>

```
rustc 1.75.0-nightly (fdaaaf9f9 2023-11-08)
binary: rustc
commit-hash: fdaaaf9f923281ab98b865259aa40fbf93d72c7a
commit-date: 2023-11-08
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.4
```

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/117761

cc @saethlin

### CODE REVIEW

File: tests/codegen/overaligned-constant.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// GVN may create indirect constants with higher alignment than their type requires. Verify that we
+// do not ICE during codegen, and that the LLVM constant has the higher alignment.
+//
+// compile-flags: -O -Zmir-enable-passes=+GVN -Cdebuginfo=2
```
Comment (saethlin) [inline tests/codegen/overaligned-constant.rs:4]: I'm wary of this combination of flags. Why do we need `-Cdebuginfo=2`? Can we use `-Zmir-opt-level=0` so this doesn't depend on which MIR opts `-O` enables?

Comment (saethlin) [reply tests/codegen/overaligned-constant.rs]: Nice. Thanks.

File: compiler/rustc_codegen_ssa/src/mir/operand.rs [hunk 2]
```diff
@@ -132,7 +132,7 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         offset: Size,
     ) -> Self {
         let alloc_align = alloc.inner().align;
-        assert_eq!(alloc_align, layout.align.abi);
+        assert!(alloc_align >= layout.align.abi);
```
Comment (wesleywiser) [inline compiler/rustc_codegen_ssa/src/mir/operand.rs:135]: Would it be more precise to say that `alloc_align` must be a multiple of `layout.align.abi`?

Comment (cjgillot) [reply compiler/rustc_codegen_ssa/src/mir/operand.rs:135]: Both are instances of `Align`, which only stores the exponent of the alignment. So the inequality does what we want.

**Review by wesleywiser — APPROVED** [review]

**Pull request merged**